### PR TITLE
Handle Option correctly

### PR DESCRIPTION
**Pull request by honnix**
# TL;DR
Handle Scala `Option` correctly by using correct type.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
For case classes like:
```scala
case class ScalarNested(
    foo: String,
    bar: Option[String],
    nestedNested: Option[ScalarNestedNested]
)
case class ScalarNestedNested(foo: String, bar: Option[String])
```
the current code doesn't handle typing correctly because calling `param.typeSignature.dealias.typeArgs.head.typeSymbol` on `Option` type does not give what we need, and it should be `param.typeSignature.dealias.typeArgs.head` instead.

Also we are not handling nested `Option` correctly, although in practice there is no good reason to define `Option[Option[String]]`.

## Tracking Issue
_NA_

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/SdkLiteralTypes.scala [hunk 1]
```diff
@@ -297,41 +297,39 @@ object SdkLiteralTypes {
...
+          value.asInstanceOf[Double].toShort
+        } else if (tpe =:= typeOf[Int]) {
+          value.asInstanceOf[Double].toInt
+        } else if (tpe =:= typeOf[Long]) {
+          value.asInstanceOf[Double].toLong
+        } else if (tpe =:= typeOf[Float]) {
+          value.asInstanceOf[Double].toFloat
+        } else if (tpe <:< typeOf[Option[Any]]) { // this has to be before Product check because Option is a Product
+          if (value == None) { // None is used to represent Struct.Value.Kind.NULL_VALUE when converting struct to map
+            None
```
Comment (honnix) [inline flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/SdkLiteralTypes.scala:313]: This is another bug fix. We have been converting things into `Some(None)`, due to type conformation test failure against `T` in `Option[T]`, which is exactly what this PR set off to fix. Bug on top of bug :(

Hopefully the extended unit tests and ITs cover better.

**Review by andresgomezfrr — APPROVED** [review]
